### PR TITLE
group names based on bucket UUID

### PR DESCRIPTION
**Pull request by tedsharpe**
Addresses DSDEEPB-1234.

- [x] **Submitter**: Rebase to master. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] **Submitter**: **Assign PR to** 
- [x] : does a once-over of the PR
- [x] : Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  * Rest of team may comments on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter updates documentation as needed
  * Submitter rebases to master again if necessary
  * Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x] **LR**: sign off, **assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to master
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: Verify swagger UI on dev server still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala [hunk 1]
```diff
@@ -56,76 +57,34 @@ class HttpGoogleCloudStorageDAO(
...
+          //You can either remove all the objects manually, or you can set up lifecycle management on the bucket.
+          //This can be used to auto-delete all objects next time the Google lifecycle manager runs (~every 24h).
+          //More info: http://bit.ly/1WCYhhf and http://bit.ly/1Py6b6O
+          val deleteEverythingRule = new Lifecycle.Rule()
+            .setAction(new Action().setType("delete"))
+            .setCondition(new Condition().setAge(0))
+          val lifecycle = new Lifecycle().setRule(List(deleteEverythingRule))
+          val bucket = buckets.get(bucketName).execute()
+          bucket.setLifecycle(lifecycle)
+
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala]: why is this change necessary?

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -135,45 +135,26 @@ class WorkspaceService(userInfo: UserInfo, dataSource: DataSource, containerDAO:
...
+        val accessLevel = gcsDAO.getMaximumAccessLevel(userInfo.userEmail,workspaceContext.workspace.bucketName)
+        if ( accessLevel < WorkspaceAccessLevel.Read )
+          RequestComplete(http.StatusCodes.NotFound, noSuchWorkspaceMessage(workspaceName))
+        else {
+          val response = WorkspaceListResponse(accessLevel,
+            workspaceContext.workspace,
+            getWorkspaceSubmissionStats(workspaceContext, txn),
+            gcsDAO.getOwners(workspaceContext.workspace.bucketName))
+          RequestComplete(response)
+        }
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: why this change?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala (hunk 1 above)
Comment (tedsharpe) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala]: It isn't.  The code is reordered for slightly improved sanity:  It isn't necessary to create the lifecycle rule and schedule deletions if you can delete the bucket.  So we try that first.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 2 above)
Comment (tedsharpe) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Because the old code was silly.  It was getting the user's access level (which it had just retrieved and checked, by the way) by getting all groups that the user was a part of and filtering that by workspace.  Now we just get the access level once.

**Comment by dvoet** [comment]
I would prefer we pass workspace around instead of the bucket name, there should be very few methods that have a parameter called bucket name and they should be doing something with a bucket. 

I think there should be a property on the workspace called workspaceId which is the UUID and the bucket name and groups names be derived from that. Lookups should be based on the workspaceId, not the bucket name. I feel like the intent would be clearer.

**Comment by tedsharpe** [comment]
Changing the Workspace case class to have a workspaceId rather than a bucketName property will change the rawls API and invalidate the DB.  Are you sure you want to do this?

**Comment by tedsharpe** [comment]
I'll change the last 20 references to Workspace.bucketName to workspaceId as soon as I get confirmation that we want to change the API and scratch the DB.

**Comment by dvoet** [comment]
I think adding workspaceId is the right thing. I don't think rawls is yet mature enough to let api changes or database invalidation enter into the discussion.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala [hunk 3]
```diff
@@ -138,127 +98,133 @@ class HttpGoogleCloudStorageDAO(
...
+    //a workspace should always have an owner, but just in case for some reason it doesn't...
+    val ownersQuery = members.list(toGroupId(getBucketName(workspaceId), WorkspaceAccessLevel.Owner)).execute()
+
+    Option(ownersQuery.getMembers).getOrElse(List.empty[Member].asJava).map(_.getEmail)
+  }
+
+  override def getMaximumAccessLevel(userId: String, workspaceId: String): WorkspaceAccessLevel = {
+    val bucketName = getBucketName(workspaceId)
+    val members = getGroupDirectory.members
+    def testLevel( levels: Seq[WorkspaceAccessLevel] ): WorkspaceAccessLevel = {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala]: add @tailrec annotation

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 4]
```diff
@@ -242,30 +223,14 @@ class WorkspaceService(userInfo: UserInfo, dataSource: DataSource, containerDAO:
...
-                    containerDAO.methodConfigurationDAO.list(sourceWorkspaceContext, txn).foreach { methodConfig =>
-                      containerDAO.methodConfigurationDAO.save(destWorkspaceContext,
-                        containerDAO.methodConfigurationDAO.get(sourceWorkspaceContext, methodConfig.namespace, methodConfig.name, txn).get, txn)
-                    }
-                    RequestCompleteWithLocation((StatusCodes.Created, destWorkspaceContext.workspace), destWorkspaceName.path)
-                }
-            }
+        withNewWorkspaceContext(WorkspaceRequest(destWorkspaceName.namespace,destWorkspaceName.name,sourceWorkspaceContext.workspace.attributes),txn) { destWorkspaceContext =>
+          containerDAO.entityDAO.cloneAllEntities(sourceWorkspaceContext, destWorkspaceContext, txn)
+          // TODO add a method for cloning all method configs, instead of doing this
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: what's up with this TODO?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala [hunk 5]
```diff
@@ -56,76 +57,35 @@ class HttpGoogleCloudStorageDAO(
...
-
-  override def createBucket(userInfo: UserInfo, projectId: String, bucketName: String): Unit = {
-    val credential = getBucketCredential(userInfo)
-    val storage = getStorage(credential)
-    val bucket =  new Bucket().setName(bucketName)
-    storage.buckets().insert(projectId, bucket).execute()
+  override def createBucket(userInfo: UserInfo, projectId: String): Try[String] = {
+    val buckets = getStorage(getBucketCredential(userInfo)).buckets
+    retry(2) { // in case of bucket name collisions
+      val workspaceId = UUID.randomUUID.toString
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala]: it is unexpected to create the workspace id within the createBucket function. Make workspaceId a parameter.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 4 above)
Comment (tedsharpe) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: It's always been there.  (used to be line 260.)

File: src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 6]
```diff
@@ -40,7 +40,7 @@ case class WorkspaceRequest (
 case class Workspace (
                       namespace: String,
                       name: String,
-                      bucketName: String,
+                      workspaceId: String,
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: I am torn here, I think we are going to need to expose the bucket name for the workspace. I originally thought we could hide the bucket but users might want to upload files.

Would it work to have a val in the Workspace object which constructs the bucket name from the workspaceId?

**Comment by dvoet** [comment]
I don't think it is worth the change here... maybe tech debt, but there is too much workspaceId in function parameters. I feel it would be more OO to pass the Workspace object around instead, especially because it looks like we have the workspace in just about all cases.

**Comment by dvoet** [comment]
re-reviewed

File: src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala (hunk 6 above)
Comment (tedsharpe) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: It is exposed via the new getBucketName(workspaceId: String) method in the DAO.

Comment (tedsharpe) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: We can tear up the last few commits and go back to bucketName in the Workspace if you'd prefer.

**Comment by tedsharpe** [comment]
Passing Workspace rather than workspaceId violates your usual dictum about making the dependencies minimal and explicit, it seems to me, but if you'd like to go that way I'd prefer to just stop the suffering right now and fix it.

**Comment by dvoet** [comment]
RE: workspaceId parameters everywhere, the dependency dictum applies to services and configuration, not to decomposing data structures.

 :+1: merge it
